imxrt-hal/src/can/frame.rs

```diff
@@ -0,0 +1,445 @@
+        Self(code)
+    }
+
+    /// Creates a new code reg for a standard identifier.
+    pub fn new_standard() -> Self {
+        Self::new(0x00)
+    }
+
+    /// Creates a new code reg for an extended identifier.
+    pub fn new_extended() -> Self {
+        Self::new(Self::IDE_MASK)
+    }
+
+    /// Returns the register value.
+    #[inline(always)]
+    pub fn to_code_reg(&self) -> u32 {
+        self.0
+    }
+
+    /// Set the 4 bit code content for a CodeReg
+    #[inline(always)]
+    pub fn set_code(&mut self, code: FlexCanMailboxCSCode) -> Self {
+        Self::new(self.0 & ((code as u32) << Self::CODE_SHIFT))
+    }
+
+    /// Get the 4 bit code content for a CodeReg.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if a matching [`FlexCanMailboxCSCode`] is not found.
+    #[inline(always)]
+    pub fn code(&self) -> FlexCanMailboxCSCode {
+        FlexCanMailboxCSCode::try_from(((self.0 & Self::CODE_MASK) >> Self::CODE_SHIFT) as u8)
+            .unwrap()
+    }
+
+    pub fn timestamp(&self) -> u16 {
+        ((self.0 & Self::TIMESTAMP_MASK) >> Self::TIMESTAMP_SHIFT) as u16
+    }
+
+    pub fn dlc(&self) -> u8 {
+        ((self.0 & Self::DLC_MASK) >> Self::DLC_SHIFT) as u8
+    }
+
+    /// Returns `true` if the code reg is an extended identifier.
+    #[inline(always)]
+    fn is_extended(self) -> bool {
+        self.0 & Self::IDE_MASK != 0
+    }
+
+    /// Returns `true` if the code reg is a standard identifier.
+    #[inline(always)]
+    fn is_standard(self) -> bool {
+        !self.is_extended()
+    }
+
+    /// Sets the remote transmission (RTR) flag. This marks the identifier as
+    /// being part of a remote frame.
+    #[must_use = "returns a new IdReg without modifying `self`"]
+    pub fn with_rtr(self, rtr: bool) -> Self {
+        if rtr {
+            Self::new(self.0 | Self::RTR_MASK)
+        } else {
+            Self::new(self.0 & !Self::RTR_MASK)
+        }
+    }
+
+    /// Returns `true` if the identifer is part of a remote frame (RTR bit set).
+    #[inline(always)]
+    pub fn rtr(self) -> bool {
+        self.0 & Self::RTR_MASK != 0
+    }
+}
+
+/// Identifier of a CAN message.
+///
+/// Can be either a standard identifier (11bit, Range: 0..0x3FF) or a
+/// extendended identifier (29bit , Range: 0..0x1FFFFFFF).
+///
+/// The `Ord` trait can be used to determine the frame’s priority this ID
+/// belongs to.
+/// Lower identifier values have a higher priority. Additionally standard frames
+/// have a higher priority than extended frames and data frames have a higher
+/// priority than remote frames.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct IdReg(u32);
+
+impl IdReg {
+    const STANDARD_SHIFT: u32 = 18;
+    const EXTENDED_SHIFT: u32 = 0;
+
+    /// Creates a new standard identifier (11bit, Range: 0..0x7FF)
+    ///
+    /// Panics for IDs outside the allowed range.
+    #[inline(always)]
+    pub fn new(id: u32) -> Self {
+        Self(id)
+    }
```

Comment:
> Panics for IDs outside the allowed range.

This constructor doesn't seem to panic. Are the docs out of date, or is this missing a bounds check?

---

What are your thoughts on making `new` a private constructor? This would prevent users from creating an invalid `IdReg`, which might be useful if we intend to have no bounds check.

Reply:
> > Panics for IDs outside the allowed range.
> 
> This constructor doesn't seem to panic. Are the docs out of date, or is this missing a bounds check?
> 

Updated to:

```rust
/// Creates a new standard identifier (11bit, Range: 0..=0x7FF)
///
/// Panics for IDs outside the allowed range.
#[inline(always)]
pub fn new_standard(id: StandardId) -> Self {
    assert!(id.as_raw() <= StandardId::MAX.as_raw());
    let id = u32::from(id.as_raw()) << Self::STANDARD_SHIFT;
    Self::new(id)
}

/// Creates a new extendended identifier (29bit , Range: 0..=0x1FFFFFFF).
///
/// Panics for IDs outside the allowed range.
#[inline(always)]
pub fn new_extended(id: ExtendedId) -> Self {
    assert!(id.as_raw() <= ExtendedId::MAX.as_raw());
    let id = u32::from(id.as_raw()) << Self::EXTENDED_SHIFT;
    Self::new(id)
}
```

> What are your thoughts on making `new` a private constructor? This would prevent users from creating an invalid `IdReg`, which might be useful if we intend to have no bounds check.

Yes, no issues making `new` private.